crates/ark/src/variables/variable.rs

```diff
@@ -918,132 +931,216 @@ impl PositronVariable {
+                Ok(i) => i,
+            };
+
+            let name_len = match parsed_access_key[2].parse::<usize>() {
+                Err(_) => return Ok(None), // Not an access_key in the required format
+                Ok(name_len) => name_len,
+            };
 
-                        _ => return Err(harp::error::Error::InspectError { path: path.clone() }),
+            let name = match parsed_access_key[3] {
+                "" => RObject::from(r_null()), // Empty string, means a `NULL` name
+                nm => {
+                    if nm.len() == name_len {
+                        RObject::from(nm)
+                    } else {
+                        // Name has been truncated, we pass it as `NULL`
+                        RObject::from(r_null())
                     }
                 },
+            };
 
-                EnvironmentVariableNode::VectorElement { .. } => {
-                    return Err(harp::error::Error::InspectError { path: path.clone() });
-                },
+            ArkGenerics::VariableGetChildAt.try_dispatch::<RObject>(object.sexp, vec![
+                RArgument::new("index", RObject::from(index + 1)), // Index is 0-based, so we convert to 1-based for R.
+                RArgument::new("name", RObject::from(name)),
+            ])
+        });
 
-                EnvironmentVariableNode::Matrixcolumn { object, index } => unsafe {
-                    let dim = IntegerVector::new(Rf_getAttrib(*object, R_DimSymbol))?;
-                    let n_row = dim.get_unchecked(0).unwrap() as isize;
+        match result {
+            Err(err) => {
+                // It's not safe to apply default methods in this case, because we rely on custom
+                // access keys, which could indicate the access index depending on the node implementation.
+                // See for instance, how it's used to index lists and atomic vectors.
+                return Err(harp::Error::Anyhow(err));
+            },
+            Ok(None) => {
+                // The object doesn't have a custom get_child_at method. We apply
+                // the default built-in methods.
+            },
+            Ok(Some(child)) => return Ok(EnvironmentVariableNode::Concrete { object: child }),
+        };
+
+        // For S4 objects, we acess child nodes using R_do_slot.
+        if object.is_s4() {
+            let name = unsafe { r_symbol!(access_key) };
+            let child: RObject =
+                harp::try_catch(|| unsafe { R_do_slot(object.sexp, name) }.into())?;
+            return Ok(EnvironmentVariableNode::Concrete { object: child });
+        }
 
-                    // TODO: use ? here, but this does not return a crate::error::Error, so
-                    //       maybe use anyhow here instead ?
-                    let row_index = path_element.parse::<isize>().unwrap();
+        // R6 objects may be accessed with special elements called <methods> and <private>.
```

Comment:
Do you think we could move all that code to the new Ark methods?

It'd be nice to move all that complexity to the R side.

Reply:
Yes, I have an initial implementation here: https://github.com/dfalbel/testVariablesPaneExtension/blob/main/R/R6.R
It's still pretty much a draft and depends a lot on rlang, but we could move it to ark and inron it out. I'd rather do it in a follow up PR though.